associations/tests/test_library.py

```diff
@@ -2,111 +2,246 @@
-
-        self.login('17admin')  # Just global administrator, not library administrator.
-        res = self.post('library/', data={'id': 'bde', 'enabled': 'true'})
-        self.assertEqual(res.status_code, 403)
-        self.assertRaises(ObjectDoesNotExist, Library.objects.get, pk='bde')
-
-        self.login('17admin_bde')  # Just association administrator, not library administrator.
-        res = self.post('library/', data={'id': 'bde', 'enabled': 'true'})
-        self.assertEqual(res.status_code, 403)
-        self.assertRaises(ObjectDoesNotExist, Library.objects.get, pk='bde')
+        for user in ['17simple',  # A simple user.
+                     '17admin',  # A global administrator.
+                     '17admin_bde']:  # An association administrator.
+            self.login(user)
+            res = self.post('library/', data={'id': 'bde', 'enabled': 'true', 'association': 'bde', 'loanables': []})
+            self.assertEqual(res.status_code, 403)
+            self.assertRaises(ObjectDoesNotExist, Library.objects.get, pk='bde')
 
     def test_if_library_admin_then_can_create_own_library(self):
         self.login('17library_bde')  # Library administrator.
-        res = self.post('library/', data={'id': 'bde', 'enabled': 'true'})
-        self.assertEqual(res.status_code, 201)
+        res = self.post('library/',
+                        data={'id': 'bde', 'enabled': 'true', 'loanables': [], 'association': 'bde'})
+        self.assertEqual(res.status_code, 201, msg=res.content)
         self.assertTrue(Library.objects.filter(pk='bde').exists())
         self.assertEqual(Association.objects.get(pk='bde').library, Library.objects.get(pk='bde'))
 
     ##########
-    # DELETE #
+    # UPDATE #
     ##########
 
-    def test_if_not_library_admin_then_cannot_delete_library(self):
-        self.login('17simple')
-        res = self.delete('library/bd-tek/')
-        self.assertEqual(res.status_code, 403)
-        self.assertTrue(Library.objects.filter(pk='bd-tek').exists())
+    def test_if_not_library_admin_then_cannot_update_library(self):
+        for user in ['17simple',  # A simple user.
+                     '17admin',  # A global administrator.
+                     '17admin_bde']:  # An association administrator.
+            self.login(user)
+            res = self.patch('library/bd-tek/', data={'enabled': 'false'})
+            self.assertEqual(res.status_code, 403)
+            self.assertTrue(Library.objects.get(pk='bd-tek').enabled)
 
-        self.login('17admin')  # Just global administrator, not library administrator.
-        res = self.delete('library/bd-tek/')
-        self.assertEqual(res.status_code, 403)
-        self.assertTrue(Library.objects.filter(pk='bd-tek').exists())
+    def test_if_library_admin_then_can_update_library(self):
+        self.login('17library_bd-tek')
+        res = self.patch('library/bd-tek/', data={'enabled': 'false'})
+        self.assertIn(res.status_code, [200, 204])
```

Comment:
Why are there 2 codes ? Isn't it deterministic ?

Reply:
I think it is, but I'm not sure about which one is sent by the server…

Reply:
Choose one, we'll fix it when needed